File: tools/pandoc-site.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE CPP #-}

import Text.Pandoc
import Text.Pandoc.Walk (walk, query)
import Text.Pandoc.Class (runPure)
import Text.Pandoc.Builder (text, toList)
import Text.Pandoc.Options (def)
import Text.Pandoc.JSON (toJSONFilter)
import Text.Pandoc.Shared (hierarchicalize, Element(..))

import Data.Either (fromRight)
import Data.Maybe (fromMaybe)
import Data.Text (unpack)

collectHeaders :: Block -> [Block]
collectHeaders header@(Header _ (_, classes, _) _) =
  if "notoc" `elem` classes
    then []
    else [header]
collectHeaders _ = []

markupLink :: Attr -> [Inline] -> Inline
markupLink (headerId, _, headerProperties) headerText
    = let linkText = fromMaybe headerText (fmap (toList . text) $ lookup "toc" headerProperties)
       in Link nullAttr linkText (("#" ++ headerId), headerId)

markupElement :: Element -> [Block]
markupElement (Sec _ _ hAttr hText headers)
  | headers == [] = [link]
  | otherwise     = [link, markupElements headers]
  where link = Plain [markupLink hAttr hText]
markupElement n = error $    "'markupElement' should only be passed a 'Sec'\n"
                          ++ "    saw: " ++ show n

markupElements :: [Element] -> Block
markupElements = OrderedList (1, Decimal, Period) . map markupElement

createTable :: [Element] -> Block
createTable [] = Null
createTable headers
    = let navBegin  = "<nav id=\"toc\" class=\"right-toc\">"
          navEnd    = "</nav>"
          tocDoc    = Pandoc nullMeta [Para [Str "Contents"], markupElements headers]
          tocString = unpack . fromRight mempty . runPure . writeHtml5String def $ tocDoc
       in RawBlock "html" (navBegin ++ "\n" ++ tocString ++ "\n" ++ navEnd)

generateTOC :: Block -> Block -> Block
generateTOC toc (Para [Str "$toc$"]) = toc
generateTOC _   x                    = x

tableOfContents :: Pandoc -> Pandoc
tableOfContents ast
    = let headers = query collectHeaders ast
          toc     = createTable . hierarchicalize $ headers
       in walk (generateTOC toc) ast

main :: IO ()
main = toJSONFilter tableOfContents

File: tools/pandoc-site.lua
function gatherHeaders(bs)
  headers = {}
  for _,b in pairs(bs) do
    if b.t == "Header" then table.insert(headers, b) end
  end
  return headers
end

function isTocBlock(blk)
  if not (blk.t               == "Para")  then return false end
  if not  blk.content[1]                  then return false end
  if not (blk.content[1].t    == "Str")   then return false end
  if not (blk.content[1].text == "$toc$") then return false end
  return true
end

function markupLink(hAttr, headerText)
  local headerId         = hAttr.identifier
  local headerProperties = hAttr.attributes
  return pandoc.Link(headerText, "#" .. headerId, headerId)
end

function markupElement(elem)
  local hAttr   = elem.attr
  local hText   = elem.label
  local hNested = elem.contents
  local link    = pandoc.Plain(markupLink(hAttr, hText))
  if not hNested[1] then return {link} end
  return {link, markupElements(hNested)}
end

function markupElements(elems)
  local newElems = {}
  for _,e in pairs(elems) do
    table.insert(newElems, markupElement(e))
  end
  return pandoc.OrderedList(newElems, {"1", "Decimal", "Period"})
end

function createTable(elems)
  local navBegin  = pandoc.RawBlock("html", "<nav id=\"toc\" class=\"right-toc\">")
  local navEnd    = pandoc.RawBlock("html", "</nav>")
  local contentsP = pandoc.Para(pandoc.Str("Contents"))
  return {navBegin, contentsP, markupElements(elems), navEnd}
end

function Pandoc(doc)
  newBlocks = {}
  tocBlocks = createTable(pandoc.utils.hierarchicalize(gatherHeaders(doc.blocks)))
  for _,blk in pairs(doc.blocks) do
    if isTocBlock(blk) then
      for _,tocBlk in pairs(tocBlocks) do
        table.insert(newBlocks, tocBlk)
      end
    else
      table.insert(newBlocks, blk)
    end
  end
  return pandoc.Doc(newBlocks, doc.meta)
end
